[PackageTool.ps1] Conversion from JSON failed with error: Unterminated string.

There's an error running `Update-ApplicationSubmission -ReplacePackages` regarding notes for certification.

```
##[error]ConvertFrom-Json : Conversion from JSON failed with error: Unterminated string. Expected delimiter: ". Path 'appSubmission.notesForCertification', line 1, position 2059.
At D:\a\MediaFileManager\MediaFileManager\broker\StoreBroker\PackageTool.ps1:3276 char:22
```

I am not making any custom package setup; just a clone of previous submission and uploading new packages with the `-ReplacePackages` flag.

## Script

Here's my Powershell script, running in a GitHub Action on a windows machine. Disregard any envVars you dont recognize, they're set earlier on the VM. 

If you want to see the rest of it, [it is public here MediaFileManager CD.yaml](https://github.com/LanceMcCarthy/MediaFileManager/blob/main/.github/workflows/cd.yml).

```
$uploadPackage = "D:\a\MediaFileManager\MediaFileManager\src\MediaFileManager\PackageProject\StoreUploadPackages\PackageProject_" + $env:UWP_VERSION + "_" + $env:UwpBundlePlatform + "_bundle.appxupload"
$username = $env:PartnerCenterClientId
$password = ConvertTo-SecureString $env:PartnerCenterClientSecret -AsPlainText -Force
$cred = New-Object System.Management.Automation.PSCredential ($username, $password)

# install StoreBroker
Set-ExecutionPolicy RemoteSigned -Force
git clone https://github.com/Microsoft/StoreBroker.git 'D:\a\MediaFileManager\MediaFileManager\broker\'
Import-Module -Force 'D:\a\MediaFileManager\MediaFileManager\broker\StoreBroker'

# Authenticating with DevCenter
Set-StoreBrokerAuthentication -TenantId $env:PartnerCenterTenantId -Credential $cred

# Generating PDP
$pdpFolderPath = [System.Environment]::GetFolderPath('Desktop')
$pdpFileName = 'Pdp.xml' 
D:\a\MediaFileManager\MediaFileManager\broker\\Extensions\ConvertFrom-ExistingSubmission.ps1 -AppId $env:PartnerCenterStoreId -Release $env:UWP_VERSION -PdpFileName $pdpFileName -OutPath $pdpFolderPath

# Generating SBConfig file
$brokerConfigFilePath = Join-Path -Path ([System.Environment]::GetFolderPath('Desktop')) -ChildPath 'SBConfig.json'
New-StoreBrokerConfigFile -AppId $env:PartnerCenterStoreId -Path $brokerConfigFilePath

# Generate submission Data package
$submissionDataFilePath = Join-Path -Path ([System.Environment]::GetFolderPath('Desktop')) -ChildPath 'submission.json'
New-SubmissionPackage -ConfigPath (Join-Path -Path ([System.Environment]::GetFolderPath('Desktop')) -ChildPath 'SBConfig.json') -PDPRootPath ([System.Environment]::GetFolderPath('Desktop')) -PDPInclude 'Pdp.xml' -ImagesRootPath 'D:\a\MediaFileManager\MediaFileManager\src\MediaFileManager\PackageProject\Images' -AppxPath $uploadPackage -OutPath $pdpFolderPath -OutName $submissionDataFilePath

# submitting app
Update-ApplicationSubmission -ReplacePackages -AppId $env:PartnerCenterStoreId -SubmissionDataPath $submissionDataFilePath -PackagePath $uploadPackage -AutoCommit -Force

```


## Output

Here is the output of the run.  To keep it short, I removed the duplicate "I'm busy" updates. 



```
Cloning into 'D:\a\MediaFileManager\MediaFileManager\broker'...

|  Elapsed: 2 second(s) [Getting access token]
DONE - Operation took 2 second(s) [Getting access token]

/  Elapsed: 2 second(s) [Getting data for AppId: 9PD3JFK7W5MB]
DONE - Operation took 2 second(s) [Getting data for AppId: 9PD3JFK7W5MB]
2020-08-11 22:34:07 : runneradmin : Telemetry is currently enabled.  It can be disabled by setting "$global:SBDisableTelemetry = $true". Refer to USAGE.md#telemetry for more information.


-  Elapsed: 4 second(s) [Retrieving nuget package: Microsoft.Bcl.Async]
DONE - Operation took 4 second(s) [Retrieving nuget package: Microsoft.Bcl.Async]
2020-08-11 22:34:13 : runneradmin : To avoid this download delay in the future, copy the following file:
  [C:\Users\RUNNER~1\AppData\Local\Temp\e229bc7e-692a-44fd-8ba5-82b48e7b62d7\Microsoft.Bcl.Async.1.0.168\lib\net40\Microsoft.Threading.Tasks.dll]
either to:
  [D:\a\MediaFileManager\MediaFileManager\broker\StoreBroker]
or to:
  a directory of your choosing, and save that directory path to $SBAlternateAssemblyDir


\  Elapsed: 2 second(s) [Retrieving nuget package: Microsoft.Diagnostics.Tracing.EventSource.Redist]
DONE - Operation took 2 second(s) [Retrieving nuget package: Microsoft.Diagnostics.Tracing.EventSource.Redist]
2020-08-11 22:34:15 : runneradmin : To avoid this download delay in the future, copy the following file:
  [C:\Users\RUNNER~1\AppData\Local\Temp\e229bc7e-692a-44fd-8ba5-82b48e7b62d7\Microsoft.Diagnostics.Tracing.EventSource.Redist.1.1.24\lib\net35\Microsoft.Diagnostics.Tracing.EventSource.dll]
either to:
  [D:\a\MediaFileManager\MediaFileManager\broker\StoreBroker]
or to:
  a directory of your choosing, and save that directory path to $SBAlternateAssemblyDir

-  Elapsed: 2 second(s) [Retrieving nuget package: Microsoft.ApplicationInsights]
DONE - Operation took 2 second(s) [Retrieving nuget package: Microsoft.ApplicationInsights]
2020-08-11 22:34:18 : runneradmin : To avoid this download delay in the future, copy the following file:
  [C:\Users\RUNNER~1\AppData\Local\Temp\e229bc7e-692a-44fd-8ba5-82b48e7b62d7\Microsoft.ApplicationInsights.2.0.1\lib\net45\Microsoft.ApplicationInsights.dll]
either to:
  [D:\a\MediaFileManager\MediaFileManager\broker\StoreBroker]
or to:
  a directory of your choosing, and save that directory path to $SBAlternateAssemblyDir

/  Elapsed: 14 second(s) [Getting data for AppId: 9PD3JFK7W5MB SubmissionId: 1152921505688863279]
DONE - Operation took 14 second(s) [Getting data for AppId: 9PD3JFK7W5MB SubmissionId: 1152921505688863279]
2020-08-11 22:34:36 : runneradmin : PDP's have been created here: C:\Users\runneradmin\Desktop
2020-08-11 22:34:36 : runneradmin : You now need to find all of your assets and place them here: <ImagesRootPath>\2020.811.75.0\<langcode>\...
  where <ImagesRootPath> is the path defined in your config file,
  and <langcode> is the same langcode for the directory of the PDP file referencing those assets.
2020-08-11 22:34:36 : runneradmin : Every language that has a PDP references the following assets:
	MFM1.png
	MFM2.png
	MFM3.png


/  Elapsed: 1 second(s) [Getting data for AppId: 9PD3JFK7W5MB]
DONE - Operation took 2 second(s) [Getting data for AppId: 9PD3JFK7W5MB]

/  Elapsed: 6 second(s) [Getting data for AppId: 9PD3JFK7W5MB SubmissionId: 1152921505688863279]
DONE - Operation took 6 second(s) [Getting data for AppId: 9PD3JFK7W5MB SubmissionId: 1152921505688863279]
Write-Error: D:\a\MediaFileManager\MediaFileManager\broker\StoreBroker\PackageTool.ps1:3725
Line |
3725 |          Write-Log -Exception $_ -Level Error
     |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | 2020-08-11 22:34:48 : runneradmin :  ConvertFrom-Json:
     | D:\a\MediaFileManager\MediaFileManager\broker\StoreBroker\PackageTool.ps1:3276 Line |
     | 3276 |      return ($lines | ConvertFrom-Json)      |                     
     | ~~~~~~~~~~~~~~~~      | Conversion from JSON failed with error: Unterminated string.
     | Expected delimiter: ". Path      | 'appSubmission.notesForCertification', line 1, position
     | 2059. 

##[error]ConvertFrom-Json : Conversion from JSON failed with error: Unterminated string. Expected delimiter: ". Path 'appSubmission.notesForCertification', line 1, position 2059.
At D:\a\MediaFileManager\MediaFileManager\broker\StoreBroker\PackageTool.ps1:3276 char:22
+     return ($lines | ConvertFrom-Json)
+                      ~~~~~~~~~~~~~~~~
+ CategoryInfo          : NotSpecified: (:) [ConvertFrom-Json], ArgumentException
+ FullyQualifiedErrorId : System.ArgumentException,Microsoft.PowerShell.Commands.ConvertFromJsonCommand
```

However, if you would rather see the entire step's output, it is also public here. [Media File Manager - Run 204743631 - Fail](https://github.com/LanceMcCarthy/MediaFileManager/runs/204743631?check_suite_focus=true). 

Be sure to expand the arrows:
![image](https://user-images.githubusercontent.com/3520532/89957380-3fb51280-dc05-11ea-8405-00ee43513e82.png)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,10 +1,21 @@
 # StoreBroker PowerShell Module
 ## Changelog
 
+## [1.19.4](https://github.com/Microsoft/StoreBroker/tree/1.19.4) - (2020/08/14)
+### Fixes:
+
+- Fixes the logic for config generation when `tag` or `notesForCertification` from Partner Center
+  had an encoded character or a URL value.  The encoded character wasn't getting re-encoded properly
+  in the generated config, and everything after the `//` in the URL was being treated as a comment.
+
+More Info: [[pr]](https://github.com/Microsoft/StoreBroker/pull/https://github.com/Microsoft/StoreBroker/pull/198) | [[cl]](https://github.com/Microsoft/StoreBroker/commit/...)
+
+Author: [**@HowardWolosky**](https://github.com/HowardWolosky)
+
 ## [1.19.3](https://github.com/Microsoft/StoreBroker/tree/1.19.3) - (2020/04/10)
 ### Fixes:
 
-+ Updated all of the Partner Center URL's to follow the new location format.
+- Updated all of the Partner Center URL's to follow the new location format.
 
 More Info: [[pr]](https://github.com/Microsoft/StoreBroker/pull/https://github.com/Microsoft/StoreBroker/pull/185) | [[cl]](https://github.com/Microsoft/StoreBroker/commit/0fbb6b9a702cf0edf52e8f5ef6d920fb859fc954)
 
@@ -13,7 +24,7 @@ Author: [**@cartwrightluke**](https://github.com/cartwrightluke)
 ## [1.19.2](https://github.com/Microsoft/StoreBroker/tree/1.19.2) - (2018/12/14)
 ### Fixes:
 
-+ Updated the logic for finding the appxbundle/appx manifests to use direct instead of relative
+- Updated the logic for finding the appxbundle/appx manifests to use direct instead of relative
   paths
 
 More Info: [[cl]](https://github.com/Microsoft/StoreBroker/commit/9f884ff367bca72c604e57f5ce9daad6b0f4b277)

--- a/StoreBroker/Helpers.ps1
+++ b/StoreBroker/Helpers.ps1
@@ -321,57 +321,6 @@ function Get-SHA512Hash
     return [System.BitConverter]::ToString($sha512.ComputeHash($utf8.GetBytes($PlainText))) -replace '-', ''
 }
 
-function Get-EscapedJsonValue
-{
-<#
-    .SYNOPSIS
-        Escapes special characters within a string for use within a JSON value.
-
-    .DESCRIPTION
-        Escapes special characters within a string for use within a JSON value.
-
-        The Git repo for this module can be found here: http://aka.ms/StoreBroker
-
-    .PARAMETER Value
-        The string that needs to be escaped
-
-    .EXAMPLE
-        Get-EscapedJsonValue -Value 'This is my "quote". Look here: c:\windows\'
-
-        Returns back the string 'This is my \"quote\". Look here: c:\\windows\\'
-
-    .OUTPUTS
-        System.String - A string with special characters escaped for use within JSON.
-
-    .NOTES
-        Normalizes newlines and carriage returns to always be \r\n.
-#>
-    [CmdletBinding()]
-    param(
-        [Parameter(
-            Mandatory,
-            ValueFromPipeline)]
-        [AllowNull()]
-        [AllowEmptyString()]
-        [string] $Value
-    )
-
-    # The syntax of -replace is a bit confusing, so it's worth a note here.
-    # The first parameter is a regular expression match pattern.  The second parameter is a replacement string.
-    # So, when we try to do "-replace '\\', '\\", that's matching a single backslash (which has to be
-    # escaped within the match regular expression as a double-backslash), and replacing it with a
-    # string containing literally two backslashes.
-    # (And as a reminder, PowerShell's escape character is actually the backtick (`) and not backslash (\).)
-
-    # \, ", <tab>
-    $escaped = $Value -replace '\\', '\\' -replace '"', '\"' -replace '\t', '\t'
-
-    # Now normalize actual CR's and LF's with their control codes.  We'll ensure all variations are uniformly formatted as \r\n
-    $escaped = $escaped -replace '\r\n', '\r\n' -replace '\r', '\r\n' -replace '\n', '\r\n'
-
-    return $escaped
-}
-
 function ConvertTo-Array
 {
 <#

--- a/StoreBroker/PackageTool.ps1
+++ b/StoreBroker/PackageTool.ps1
@@ -125,16 +125,16 @@ function Get-StoreBrokerConfigFileContentForIapId
         $updated = $updated -replace '"lifetime": ".*",', "`"lifetime`": `"$($sub.lifetime)`","
         $updated = $updated -replace '"contentType": ".*",', "`"contentType`": `"$($sub.contentType)`","
 
-        $tag = Get-EscapedJsonValue -Value $sub.tag
-        $updated = $updated -replace '"tag": ""', "`"tag`": `"$tag`""
+        $tag = ConvertTo-Json -InputObject ($sub.tag -replace '//', '\\') # Ensure // doesn't get stripped out as a comment later on.
+        $updated = $updated -replace '"tag": ""', "`"tag`": $tag"
 
         $keywords = $sub.keywords | ConvertTo-Json -Depth $script:jsonConversionDepth
         if ($null -eq $keywords) { $keywords = "[ ]" }
         $updated = $updated -replace '(\s+)"keywords": \[.*(\r|\n)+\s*\]', "`$1`"keywords`": $keywords"
 
         # NOTES FOR CERTIFICATION
-        $notesForCertification = Get-EscapedJsonValue -Value $sub.notesForCertification
-        $updated = $updated -replace '"notesForCertification": ""', "`"notesForCertification`": `"$notesForCertification`""
+        $notesForCertification = ConvertTo-Json -InputObject ($sub.notesForCertification -replace '//', '\\') # Ensure // doesn't get stripped out as a comment later on.
+        $updated = $updated -replace '"notesForCertification": ""', "`"notesForCertification`": $notesForCertification"
 
         return $updated
     }
@@ -361,8 +361,8 @@ function Get-StoreBrokerConfigFileContentForAppId
         }
 
         # NOTES FOR CERTIFICATION
-        $notesForCertification = Get-EscapedJsonValue -Value $sub.notesForCertification
-        $updated = $updated -replace '"notesForCertification": ""', "`"notesForCertification`": `"$notesForCertification`""
+        $notesForCertification = ConvertTo-Json -InputObject ($sub.notesForCertification -replace '//', '\\') # Ensure // doesn't get stripped out as a comment later on.
+        $updated = $updated -replace '"notesForCertification": ""', "`"notesForCertification`": $notesForCertification"
 
         return $updated
     }

--- a/StoreBroker/StoreBroker.psd1
+++ b/StoreBroker/StoreBroker.psd1
@@ -6,7 +6,7 @@
     CompanyName = 'Microsoft Corporation'
     Copyright = 'Copyright (C) Microsoft Corporation.  All rights reserved.'
 
-    ModuleVersion = '1.19.3'
+    ModuleVersion = '1.19.4'
     Description = 'Provides command-line access to the Windows Store Submission REST API.'
 
     RootModule = 'StoreIngestionApi'

